### DISCUSSION TITLE
Android build issues: Builds locally but not on codemagic. Strange build command used for universal apk

### DISCUSSION THREAD

**Discussion by emmett-deen** [codemagic-ci-cd/codemagic-docs#1697] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

1. Running a build locally with matching command line arguments works fine. The commands we use locally are `flutter build apk` & `flutter build appbundle`
2. On Codemagic during the Android build phase it successfully builds with the `flutter build appbundle` command but then fails when it runs a different `android-app-bundle build-universal-apk` command.

It fails with the error
```
Failed to sign artifact file_cp_resources.apk. Failed to sign apk: Exception in thread "main" com.android.apksig.apk.MinSdkVersionException: Failed to determine APK's minimum supported platform version. Use --min-sdk-version to override
	at com.android.apksigner.ApkSignerTool.sign(ApkSignerTool.java:401)
	at com.android.apksigner.ApkSignerTool.main(ApkSignerTool.java:92)
Caused by: com.android.apksig.apk.MinSdkVersionException: Failed to determine APK's minimum supported Android platform version
	at com.android.apksig.ApkSigner.getMinSdkVersionFromApk(ApkSigner.java:968)
	at com.android.apksig.ApkSigner.sign(ApkSigner.java:287)
	at com.android.apksig.ApkSigner.sign(ApkSigner.java:223)
	at com.android.apksigner.ApkSignerTool.sign(ApkSignerTool.java:395)
	... 1 more
Caused by: com.android.apksig.apk.ApkFormatException: Missing AndroidManifest.xml
	at com.android.apksig.ApkSigner.getAndroidManifestFromApk(ApkSigner.java:924)
	at com.android.apksig.ApkSigner.getMinSdkVersionFromApk(ApkSigner.java:966)
	... 4 more
```

### Expected results

I would expect when selecting to build with both apk & appbundle it would use the flutter build command for both and both succeed.

### Actual results

Uses android build-universal-apk command with appbundle as an argument and fails.

### Build id (optional)

646b6e828e0743273d5e4762

**Comment by dtrdic** [comment]
Hey @emmett-deen, the command `flutter build appbundle --release` builds the Android App Bundle (AAB) for your Flutter project in release mode.
And `android-app-bundle build-universal-apk --bundle` is a custom command specific to your build setup on Codemagic and is used to generate a universal APK file from an Android App Bundle (AAB) file produced in the previous step. Those commands are generated as expected.

The issue here is different, let's try to look at your app first. The error is suggesting `Failed to determine APK's minimum supported Android platform version`. Can you check if you have defined `minSdkVersion` in your android/app/build.gradle file?

**Reply by emmett-deen** [reply]
It is specified. I have copied our build.gradle below. the minSDKVersion is listed under defaultConfig

```
def localProperties = new Properties()
def localPropertiesFile = rootProject.file('local.properties')
if (localPropertiesFile.exists()) {
    localPropertiesFile.withReader('UTF-8') { reader ->
        localProperties.load(reader)
    }
}

def flutterRoot = localProperties.getProperty('flutter.sdk')
if (flutterRoot == null) {
    throw new GradleException("Flutter SDK not found. Define location with flutter.sdk in the local.properties file.")
}

def flutterVersionCode = localProperties.getProperty('flutter.versionCode')
if (flutterVersionCode == null) {
    flutterVersionCode = '1'
}

def flutterVersionName = localProperties.getProperty('flutter.versionName')
if (flutterVersionName == null) {
    flutterVersionName = '1.0'
}

 def keystoreProperties = new Properties()
   def keystorePropertiesFile = rootProject.file('key.properties')
   if (keystorePropertiesFile.exists()) {
       keystoreProperties.load(new FileInputStream(keystorePropertiesFile))
   }


apply plugin: 'com.android.application'
// START: FlutterFire Configuration
apply plugin: 'com.google.gms.google-services'
// END: FlutterFire Configuration
apply plugin: 'kotlin-android'
apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"

android {
    compileSdkVersion flutter.compileSdkVersion
    ndkVersion flutter.ndkVersion

    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }

    kotlinOptions {
        jvmTarget = '1.8'
    }

    sourceSets {
        main.java.srcDirs += 'src/main/kotlin'
    }

    defaultConfig {
        applicationId "com.blackairplane.tsoa_mobile"
        // You can update the following values to match your application needs.
        // For more information, see: https://docs.flutter.dev/deployment/android#reviewing-the-build-configuration.
        minSdkVersion 21
        targetSdkVersion 31
        versionCode flutterVersionCode.toInteger()
        versionName flutterVersionName
    }
    
    signingConfigs {
        release {
            if (System.getenv()["CI"]) { // CI=true is exported by Codemagic
                storeFile file(System.getenv()["CM_KEYSTORE_PATH"])
                storePassword System.getenv()["CM_KEYSTORE_PASSWORD"]
                keyAlias System.getenv()["CM_KEY_ALIAS"]
                keyPassword System.getenv()["CM_KEY_PASSWORD"]
            } else {
                keyAlias keystoreProperties['keyAlias']
                keyPassword keystoreProperties['keyPassword']
                storeFile keystoreProperties['storeFile'] ? file(keystoreProperties['storeFile']) : null
                storePassword keystoreProperties['storePassword']
            }
        }
    }

    buildTypes {
        release {
            signingConfig signingConfigs.release
        }
    }
}

flutter {
    source '../..'
}

dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
}

```

**Reply by dtrdic** [reply]
hey @emmett-deen try setting `minSdkVersion 22`

**Reply by emmett-deen** [reply]
@dtrdic we updated the minSdkVersion and still received the same error. Here is the build ID for the build that was ran with the change `6470c70544e9da33d9a653dd`

**Reply by emmett-deen** [reply, marked as answer]
@dtrdic it seems to come from using this package: https://pub.dev/packages/file_preview

Removing it from the pubspec.yaml allows the build to work. I know you don't have any time to dig through individual packages, but if you see this and think of how a singular package may cause this error let me know! Thanks

**Reply by dtrdic** [reply]
hey @emmett-deen great job! it may seem unusual for a single package to cause such an error but this things happen. It is possible that the package's dependencies or configurations conflicted with the build process, leading to the missing AndroidManifest.xml file or other issues like this.
But the build passes, that’s the is important :) cudos
